gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -2032,9 +2031,45 @@ public GoogleCloudStorageItemInfo composeObjects(
   <RequestT extends StorageRequest<?>> RequestT configureRequest(
       RequestT request, String bucketName) {
     setRequesterPaysProject(request, bucketName);
+
+    if (request instanceof Storage.Objects.Get || request instanceof Storage.Objects.Insert) {
+      setEncryptionHeaders(request);
+    }
+
+    if (request instanceof Storage.Objects.Rewrite || request instanceof Storage.Objects.Copy) {
+      setEncryptionHeaders(request);
+      setDecryptionHeaders(request);
+    }
+
     return request;
   }
 
+  private <RequestT extends StorageRequest<?>> void setEncryptionHeaders(RequestT request) {
+    if (storageOptions.getEncryptionAlgorithm() != null
```

Comment:
With validation in options class, will it be sufficient to just check if `getEncryptionAlgorithm` is set here?

Reply:
That's logically true and makes sense for us with context, but I guess for someone who just reads this code, I think this will make more sense.

Reply:
You may consider using `checkNotNull` wrappers to the other values, when setting them instead checking in if condition, because it could hide validation bug in the options class:

```java
headers.set("x-goog-encryption-key", checkNotNull(storageOptions.getEncryptionKey()));
headers.set("x-goog-encryption-key-sha256", checkNotNull(storageOptions.getEncryptionKeyHash()));
```